url86e9087e81_JohnPease_557_HW3_tgz-pJ8-uwm_cs557_Sample4J8/src/uwm_cs557/MiniDB.java

```diff
@@ -0,0 +1,229 @@
+package uwm_cs557;
+
+import java.io.*;
+
+public class MiniDB{
+
+  private String dbName;
+
+  // simiulates the disk space manager and disk drive
+  public byte dbBlocks [][];
+  public boolean dbBlockUsed[];
+
+  /** 
+   * @return the name of the operating system files the database is stored in 
+   */
+  public String osFilename(){ return dbName + ".minidb"; }
+
+  public static BufMgr bufMgr;
+  public BufMgr getBufMgr(){ return bufMgr; }
+
+  public MiniDB_Catalog catalog;
+  public MiniDB_Catalog getCatalog(){ return catalog; }
+
+  /**
+   * startup a database with name dbName
+   */
+  public MiniDB(String dbName){
+    this.dbName = dbName;
+    this.bufMgr = new BufMgr(this);
+
+    try{
+      
+      if( MiniDB_Util.fileOrDirectoryExists( osFilename() ) ){
+	// load DB from file
+	restoreDBFromOSFile();
+	
+	MiniDB_Block catHeadBlock = bufMgr.pin_block( MiniDB_Constants.CATALOG_BLOCK_ID );
+	this.catalog = MiniDB_Catalog.createFromExisting( catHeadBlock, this );
+      }
+      else{
+	// create DB from scratch
+	this.dbBlocks = new byte[ MiniDB_Constants.DB_MAX_BLOCKS ][ MiniDB_Constants.BLOCK_SIZE ];
+	this.dbBlockUsed = new boolean[ MiniDB_Constants.DB_MAX_BLOCKS ];
+
+	for (int i=0; i<MiniDB_Constants.DB_MAX_BLOCKS; i++){
+	  for (int j=0; j< MiniDB_Constants.BLOCK_SIZE ; j++){
+	    this.dbBlocks[i][j] = (byte)0xaa;
+	  }
+	}
+	
+	// create the catalog
+	int blockID = addBlockToDB();
+	assert( blockID == MiniDB_Constants.CATALOG_BLOCK_ID );
+	MiniDB_Block catHeadBlock = bufMgr.pin_block( MiniDB_Constants.CATALOG_BLOCK_ID );
+	this.catalog = MiniDB_Catalog.createNewCatalog(catHeadBlock, this);
+	bufMgr.markDirty( catHeadBlock.blockID );
+      }
+      
+    }
+    catch(MiniDB_Exception ex){
+      // bad news
+      System.out.println( "Exception " + ex );
+      ex.printStackTrace();
+      System.exit(0);
+    }
+
+  }
+
+  private void restoreDBFromOSFile(){
+    try {
+      FileInputStream fin = new FileInputStream(osFilename());
+      ObjectInputStream ois = new ObjectInputStream(fin);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *ois*. The resource is referenced at lines 72, 73, 74. The resource is closed at line 75. There are other execution paths that do not contain closure statements, for example, when ObjectInputStream.readObject throws an exception. Ensure *ois* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).